libtenzir/builtins/operators/throttle.cpp

```diff
@@ -0,0 +1,170 @@
+      auto split_position = static_cast<size_t>(budget);
+      auto head = chunk::make_empty();
+      auto tail = chunk::make_empty();
+      auto head_offset = size_t{0};
+      std::tie(head, tail) = split_chunk(bytes, head_offset, split_position);
+      budget -= static_cast<double>(head->size());
+      head_offset += head->size();
+      co_yield std::move(head);
+      // If we didn't have enough budget to send everything in one go,
+      // send the remainder in intervals according to the configured
+      // window.
+      while (tail->size() > 0) {
+        budget = 0;
+        ctrl.set_waiting(true);
+        ctrl.self()
+          .request(alarm_clock, caf::infinite,
+                   duration_cast<caf::timespan>(window_))
+          .then(
+            [&]() {
+              ctrl.set_waiting(false);
+            },
+            [&ctrl](const caf::error& err) {
+              diagnostic::error("throttle operator failed to delay")
+                .note("encountered error: {}", err)
+                .emit(ctrl.diagnostics());
+            });
+        std::tie(head, tail) = split_chunk(
+          bytes, head_offset, static_cast<size_t>(bytes_per_window));
+        head_offset += head->size();
+        co_yield {}; // Await the alarm clock.
+        co_yield std::move(head);
+      }
+      last_timestamp = std::chrono::steady_clock::now();
+    }
+    co_return;
+  }
+
+  auto name() const -> std::string override {
+    return "throttle";
+  }
+
+  auto optimize(expression const& filter,
+                event_order order) const -> optimize_result override {
+    (void)filter, (void)order;
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, throttle_operator& x) -> bool {
+    return f.apply(x.bandwidth_per_second_);
+  }
+
+private:
+  double bandwidth_per_second_ = 0.0; // bytes
+  float_seconds window_ = {};
+};
+
+class throttle_plugin final : virtual public operator_plugin<throttle_operator>,
+                              operator_factory_plugin {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto const* docs = "https://docs.tenzir.com/operators/throttle";
+    auto parser = argument_parser{"throttle", docs};
+    auto bandwidth = located<uint64_t>{};
+    // TODO: Add option to set window size.
+    auto window = float_seconds{1};
+    parser.add(bandwidth, "<bandwidth>");
+    parser.parse(p);
+    if (bandwidth.inner == 0) {
+      diagnostic::error("`bandwidth` must be a positive number")
+        .primary(bandwidth.source)
+        .note("the unit of measurement is bytes/second")
+        .throw_();
+    }
+    return std::make_unique<throttle_operator>(bandwidth.inner, window);
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    auto bandwidth = located<uint64_t>{};
```

Comment:
Should this accept doubles?

Reply:
Ah nvm, that doesn't really make sense now that I think about it.

Reply:
In theory yes, but this seems to break the Gi/Mi/ki suffixes, which I think will be more useful in practice:

```
benno@bourbaki:~/src/tenzir/health-metrics/build$ ./bin/tenzir --tql2 'throttle 34.2Gi'
error: could not parse scalar
 --> <input>:1:10
  |
1 | throttle 34.2Gi
  |          ^^^^^^ 
  |
  = note: scalar parsing still is very rudimentary
```